src/en/news/blog/2023/reef-osds-per-nvme/index.md

```diff
@@ -0,0 +1,137 @@
+
+![](images/Ceph_Reef_-_FIO_4KB_2_OSD_NVMe_Average_Latency_Advantage.svg)
+
+The average latency reduction follows a similar pattern to what we saw with IOPS.  Average latency is slightly worse (higher) at low CPU thread counts with 2 OSDs per NVMe, but significantly better (lower) at very high CPU thread counts.
+
+# 4KB RBD Random Read and Write 99% Tail Latency
+
+![](images/Ceph_Reef_-_FIO_4KB_Random_Read_99_Tail_Lantecy.svg)
+![](images/Ceph_Reef_-_FIO_4KB_Random_Write_99_Tail_Lantecy.svg)
+
+So far we haven't seen strong evidence that deploying multiple OSDs per NVMe is a significant advantage except at very high CPU thread ratios.  There is one case though where we still see a significant advantage.  Running multiple OSDs per NVMe consistently reduces tail latency in almost all cases.
+
+![](images/Ceph_Reef_-_FIO_4KB_2_OSD_NVMe_99_Tail_Latency_Advantage.svg)
+
+The improvement in 99% Tail latency is dramatic, in some cases running 2 OSDs per NVMe reduces 99% latency to half.  While not shown here, the improvement for 99.9% latency is even better.  While for most Reef deployments running with multiple OSDs per NVMe may not show a signicant performance advantage, it may show a significant tail latency advantage.  But...Is there a cost?
+
+# 4KB RBD Random Read and Write CPU Usage
+
+![](images/Ceph_Reef_-_FIO_4KB_Random_Read_CPU_Threads_Used.svg)
+![](images/Ceph_Reef_-_FIO_4KB_Random_Write_CPU_Threads_Used.svg)
+![](images/Ceph_Reef_-_FIO_4KB_2_OSD_NVMe_CPU_Usage_Increase.svg)
+
+Despite restricting OSDs to only using a certain number of CPU threads via numactl, the 2 OSD per NVMe configuration consistently used more CPU than the 1 OSD per NVMe case.  Let's look further at how this translates into CPU Efficiency.
+
+![](images/Ceph_Reef_-_FIO_4KB_Random_Read_IOPS_CPU_Thread_Used.svg)
+![](images/Ceph_Reef_-_FIO_4KB_Random_Write_IOPS_CPU_Thread_Used.svg)
+![](images/Ceph_Reef_-_FIO_4KB_2_OSD_NVMe_IOPS_CPU_Advantage.svg)
+
+We saw previously that at a given CPU thread count, the 1 and 2 OSD per NVMe configurations performed similarly up to around 16 CPU Threads.  After that, the 2 NVMe per OSD configuration continued to scale while the single OSD configuration topped out.  We also saw that 2 OSD per NVMe configurations had significantly lower tail latency.  Here we are seeing that while that to acheive those advantages, each CPU thread is working harder in the 2 NVMe per OSD configuration to achieve the same IOPS.  That potentially could lead to higher power consumption and greater heat.  One note: The random write results reported here are being reported with replication factored in to match the results we published last fall for Pacific [here](https://ceph.io/en/news/blog/2022/ceph-osd-cpu-scaling/).  While the test configuration is not exactly the same as it was last fall, it appears we are acheiving a moderate efficiency improvement in these Reef tests.
```

Comment:
achieve
Think of Snoopy with his mouth harp:  "I before E except after C"

Reply:
yep, just a typo. :)